File: server/src/main/java/com/antigravity/converters/ArduinoConfigConverter.java
package com.antigravity.converters;

import com.antigravity.proto.ColorOrder;
import com.antigravity.proto.LedType;
import com.antigravity.proto.VoltageConfig;
import com.antigravity.protocols.arduino.ArduinoConfig;
import com.antigravity.protocols.arduino.LedString;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ArduinoConfigConverter {

  public static ArduinoConfig fromProto(
      com.antigravity.proto.ArduinoConfig proto) { // fqn-collision
    Map<String, Integer> voltageConfigs = new HashMap<>();
    for (VoltageConfig vc : proto.getVoltageConfigsList()) {
      voltageConfigs.put(String.valueOf(vc.getLane()), vc.getMaxVoltage());
    }

    List<LedString> ledStrings =
        proto.getLedStringsList().stream()
            .map(
                ls ->
                    new LedString(
                        ls.getPin(),
                        new ArrayList<>(ls.getLedsList()),
                        ls.getBrightness(),
                        ls.getLedTypeValue(),
                        ls.getColorOrderValue(),
                        ls.getFlagFlashRate(),
                        new ArrayList<>(ls.getLedLaneColorOverridesList())))
            .collect(Collectors.toList());

    return new ArduinoConfig(
        proto.getName(),
        proto.getCommPort(),
        proto.getBaudRate(),
        proto.getDebounceUs(),
        proto.getHardwareType(),
        proto.getNormallyClosedLaneSensors(),
        proto.getNormallyClosedRelays(),
        proto.getGlobalInvertLights(),
        proto.getUsePitsAsLaps(),
        proto.getUseLapsForSegments(),
        ArduinoConfig.LapPinPitBehavior.fromValue(proto.getLapPinPitBehaviorValue()),
        new ArrayList<>(proto.getDigitalIdsList()),
        new ArrayList<>(proto.getAnalogIdsList()),
        ledStrings,
        voltageConfigs);
  }

  public static com.antigravity.proto.ArduinoConfig toProto(ArduinoConfig config) { // fqn-collision
    if (config == null) {
      return com.antigravity.proto.ArduinoConfig.getDefaultInstance(); // fqn-collision
    }

    com.antigravity.proto.ArduinoConfig.Builder builder = // fqn-collision
        com.antigravity.proto.ArduinoConfig.newBuilder() // fqn-collision
            .setName(config.name != null ? config.name : "")
            .setCommPort(config.commPort != null ? config.commPort : "")
            .setBaudRate(config.baudRate)
            .setDebounceUs(config.debounceUs)
            .setNormallyClosedLaneSensors(config.normallyClosedLaneSensors)
            .setNormallyClosedRelays(config.normallyClosedRelays)
            .setGlobalInvertLights(config.globalInvertLights)
            .setUsePitsAsLaps(config.usePitsAsLaps)
            .setUseLapsForSegments(config.useLapsForSegments)
            .setHardwareType(config.hardwareType);

    if (config.digitalIds != null) {
      builder.addAllDigitalIds(config.digitalIds);
    }
    if (config.analogIds != null) {
      builder.addAllAnalogIds(config.analogIds);
    }
    if (config.lapPinPitBehavior != null) {
      builder.setLapPinPitBehaviorValue(config.lapPinPitBehavior.getValue());
    }

    if (config.ledStrings != null) {
      for (LedString ls : config.ledStrings) {
        com.antigravity.proto.LedString.Builder lsBuilder = // fqn-collision
            com.antigravity.proto.LedString.newBuilder() // fqn-collision
                .setPin(ls.pin)
                .setNumUsedLeds(ls.numUsedLeds)
                .setAddressableLeds(ls.addressableLeds)
                .setBrightness(ls.brightness)
                .setLedType(LedType.forNumber(ls.ledType))
                .setColorOrder(ColorOrder.forNumber(ls.colorOrder))
                .setFlagFlashRate(ls.flagFlashRate);

        if (ls.leds != null) {
          lsBuilder.addAllLeds(ls.leds);
        }
        if (ls.ledLaneColorOverrides != null) {
          lsBuilder.addAllLedLaneColorOverrides(ls.ledLaneColorOverrides);
        }
        builder.addLedStrings(lsBuilder.build());
      }
    }

    if (config.voltageConfigs != null) {
      for (Map.Entry<String, Integer> entry : config.voltageConfigs.entrySet()) {
        try {
          int lane = Integer.parseInt(entry.getKey());
          builder.addVoltageConfigs(
              VoltageConfig.newBuilder().setLane(lane).setMaxVoltage(entry.getValue()).build());
        } catch (NumberFormatException e) {
          // Skip invalid lane keys
        }
      }
    }

    return builder.build();
  }
}

File: client/src/app/converters/arduino_config.converter.ts
import { ArduinoConfig, LedString } from "src/app/models/track";
import { com } from "src/app/proto/message";

export class ArduinoConfigConverter {
  static fromProto(proto: com.antigravity.IArduinoConfig): ArduinoConfig {
    const p = proto as any;

    const ledStrings: LedString[] = (p.ledStrings || []).map((ls: any) => ({
      pin: ls.pin || 0,
      leds: ls.leds || [],
      numUsedLeds: ls.numUsedLeds || ls.num_used_leds || 0,
      addressableLeds: ls.addressableLeds || ls.addressable_leds || 0,
      brightness: ls.brightness || 0,
      ledType: ls.ledType || ls.led_type || 0,
      flagFlashRate: ls.flagFlashRate || ls.flag_flash_rate || 0,
      ledLaneColorOverrides:
        ls.ledLaneColorOverrides || ls.led_lane_color_overrides || [],
    }));

    const voltageConfigs: { [lane: number]: number } = {};
    if (p.voltageConfigs) {
      p.voltageConfigs.forEach((vc: any) => {
        voltageConfigs[vc.lane] = vc.maxVoltage || vc.max_voltage || 0;
      });
    }

    return {
      name: p.name || "",
      commPort: p.commPort || p.comm_port || "",
      baudRate: p.baudRate || p.baud_rate || 115200,
      debounceUs: p.debounceUs || p.debounce_us || 0,
      hardwareType: p.hardwareType || p.hardware_type || 0,
      normallyClosedLaneSensors:
        p.normallyClosedLaneSensors ?? p.normally_closed_lane_sensors ?? false,
      normallyClosedRelays:
        p.normallyClosedRelays ?? p.normally_closed_relays ?? false,
      globalInvertLights: p.globalInvertLights || p.global_invert_lights || 0,
      usePitsAsLaps: p.usePitsAsLaps ?? p.use_pits_as_laps ?? false,
      useLapsForSegments:
        p.useLapsForSegments ?? p.use_laps_for_segments ?? false,
      lapPinPitBehavior: p.lapPinPitBehavior || p.lap_pin_pit_behavior || 0,
      digitalIds: p.digitalIds || p.digital_ids || [],
      analogIds: p.analogIds || p.analog_ids || [],
      ledStrings: ledStrings,
      voltageConfigs: voltageConfigs,
      // Defaulting missing fields that are in model but not in proto
      useLapsForPits: p.useLapsForPits || 0,
      useLapsForPitEnd: p.useLapsForPitEnd || 0,
    };
  }
}
